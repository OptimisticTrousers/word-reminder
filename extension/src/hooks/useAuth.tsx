// import { useMutation } from "@tanstack/react-query";
// import { useEffect, useState } from "react";
// import { useNavigate } from "react-router-dom";
// import { toast } from "react-toastify";

// const useAuth = () => {
//     const navigate = useNavigate();
//     // const [isAuthenticated, setIsAuthenticated] = useState(false);

//     // const {data, status, error, mutuate} = useMutation({mutationFn: () => {
//     //     return fetch(`${import.meta.env.VITE_API_DOMAIN}/current`)
//     // }})

//     const { data: loginData, status: loginStatus, error: loginError, mutate: loginMutate } = useMutation({
//         mutationFn: (formData) => {
//             return fetch(`${import.meta.env.VITE_API_DOMAIN}/login`, formData)
//         }, onSuccess: () => {
//             toast.success("You have successfully logged in!");
//             navigate("/")
//         }, onError: () => {
//             toast.error("There was an issue logging in!")
//         }
//     })

//     const { data: registerData, status: registerStatus, error: registerError, mutate: registerMutate } = useMutation({
//         mutationFn: (formData) => {
//             return fetch(`${import.meta.env.VITE_API_DOMAIN}/register`, formData)
//         }
//         , onSuccess: () => {
//             toast.success("You have successfully registered!");
//             navigate("/")
//         }, onError: () => {
//             toast.error("There was an issue registering!")
//         }
//     })

//     const { data: logoutData, status: logoutStatus, error: logoutError, mutate: logoutMutate } = useMutation({
//         mutationFn: (formData) => {
//             return fetch(`${import.meta.env.VITE_API_DOMAIN}/logout`, formData)
//         }
//         , onSuccess: () => {
//             toast.success("You have successfully logged out!");
//             navigate("/")
//         }, onError: () => {
//             toast.error("There was an issue logging out!")
//         }
//     })


//     useEffect(() => {

//     }, [])

//     return {
//         login: {
//             data: loginData,
//             status: loginStatus,
//             error: loginError,
//             mutate: loginMutate
//         }, register: {
//             data: registerData,
//             status: registerStatus,
//             error: registerError,
//             mutate: registerMutate
//         }, logout: {
//             data: logoutData, status: logoutStatus, error: logoutError, mutuate: logoutMutate
//         }
//     };
// }

// export default useAuth;