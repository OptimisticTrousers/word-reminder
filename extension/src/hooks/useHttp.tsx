type HttpMethods = "GET" | "POST" | "PUT" | "DELETE";

const useHttp = () => {
  const request = async (
    url: string,
    method: HttpMethods,
    options: RequestInit,
    body?: unknown
  ) => {
    const response = await fetch(url, {
      method,
      mode: "cors",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      ...options,
      body: JSON.stringify(body),
    });

    const json = await response.json();

    return json;
  };

  const get = async (url: string, options: RequestInit = {}) => {
    return request(url, "GET", options);
  };

  const post = async (
    url: string,
    body: unknown,
    options: RequestInit = {}
  ) => {
    return request(url, "POST", options, body);
  };

  const put = async (url: string, body: unknown, options: RequestInit = {}) => {
    return request(url, "PUT", options, body);
  };

  const remove = async (url: string, options: RequestInit = {}) => {
    return request(url, "DELETE", options);
  };

  return { get, post, put, remove };
};

export default useHttp;
