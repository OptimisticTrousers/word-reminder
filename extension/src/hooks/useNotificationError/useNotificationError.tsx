import { useContext } from "react";
import { useNavigate } from "react-router-dom";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../context/Notification";
import { ErrorResponse } from "../../types";
import { AUTH_NOTIFICATION_MSGS } from "../../pages/Auth/constants";

export function useNotificationError() {
  const navigate = useNavigate();
  const { showNotification } = useContext(NotificationContext);

  function showNotificationError(response: ErrorResponse) {
    const message = response.json.message;
    if (message === "Incorrect email." || message === "Incorrect password.") {
      showNotification(
        NOTIFICATION_ACTIONS.ERROR,
        AUTH_NOTIFICATION_MSGS.incorrectCredentials()
      );
    } else if (message == "User is unauthenticated.") {
      showNotification(
        NOTIFICATION_ACTIONS.ERROR,
        AUTH_NOTIFICATION_MSGS.credentialsExpired()
      );
      navigate("/login");
    } else {
      showNotification(NOTIFICATION_ACTIONS.ERROR, response.json.message);
    }
  }

  return { showNotificationError };
}
