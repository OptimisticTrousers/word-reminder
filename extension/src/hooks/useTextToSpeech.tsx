import { useState, useEffect } from "react";

const useTextToSpeech = (text: string) => {
  const [utterance, setUtterance] = useState<SpeechSynthesisUtterance | null>(
    null
  );

  useEffect(() => {
    const synth = window.speechSynthesis;
    const u = new SpeechSynthesisUtterance(text);

    setUtterance(u);

    return () => {
      synth.cancel();
    };
  }, [text]);

  const handlePlay = () => {
    const synth = window.speechSynthesis;

    synth.speak(utterance!);
  };

  return { handlePlay };
};

export default useTextToSpeech;
