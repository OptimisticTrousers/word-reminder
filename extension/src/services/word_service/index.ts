export { wordService } from "./word_service";
