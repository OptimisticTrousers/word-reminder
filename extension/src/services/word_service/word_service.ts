import { service, Params } from "../service";

export const wordService = (function (service) {
  const { get, post, remove, VITE_API_DOMAIN } = service;

  function getWordList(userId: string | undefined, params: Params) {
    return get({
      url: `${VITE_API_DOMAIN}/users/${userId}/words`,
      params,
      options: { credentials: "include" },
    });
  }

  function createWord({
    userId,
    formData,
  }: {
    userId: string | undefined;
    formData: FormData;
  }) {
    return post({
      url: `${VITE_API_DOMAIN}/users/${userId}/words`,
      options: {
        body: formData,
        credentials: "include",
      },
    });
  }

  function deleteUserWord({
    userId,
    wordId,
  }: {
    userId: string;
    wordId: string;
  }) {
    return remove({
      url: `${VITE_API_DOMAIN}/users/${userId}/${wordId}`,
      options: { credentials: "include" },
    });
  }

  return { getWordList, createWord, deleteUserWord };
})(service);
