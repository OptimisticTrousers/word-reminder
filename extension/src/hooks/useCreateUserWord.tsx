import { useMutation } from "@tanstack/react-query";
import useHttp from "./useHttp";

const useCreateUserWord = () => {
  

  return {
    createdUserWord,
    createdUserWordStatus,
    createdUserWordError,
    createdUserWordMutate,
  };
};

export default useCreateUserWord;
