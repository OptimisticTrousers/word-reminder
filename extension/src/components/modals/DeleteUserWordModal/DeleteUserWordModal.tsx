import { User } from "common";
import { useContext } from "react";
import CSSModules from "react-css-modules";
import { useOutletContext } from "react-router-dom";
import { useMutation } from "@tanstack/react-query";

import {
  NotificationContext,
  NOTIFICATION_ACTIONS,
} from "../../../context/Notification";
import { useNotificationError } from "../../../hooks/useNotificationError";
import { ModalContainer } from "../ModalContainer";
import { ErrorResponse } from "../../../types";
import { wordService } from "../../../services/word_service";
import styles from "./DeleteUserWordModal.module.css";

interface Props {
  toggleModal: () => void;
  wordId: string;
}

export const DeleteUserWordModal = CSSModules(
  function ({ toggleModal, wordId }: Props) {
    const { showNotification } = useContext(NotificationContext);
    const { showNotificationError } = useNotificationError();
    const { isPending, mutate } = useMutation({
      mutationFn: wordService.deleteUserWord,
      onSuccess: () => {
        showNotification(
          NOTIFICATION_ACTIONS.SUCCESS,
          USER_WORD_NOTIFICATION_MSGS.deleteUserWord()
        );
      },
      onError: (response: ErrorResponse) => {
        showNotificationError(response);
      },
    });
    const { user }: { user: User } = useOutletContext();
    const userId = user.id;

    function handleDelete() {
      mutate({ userId, wordId });
    }

    function handleCancel() {
      toggleModal();
    }

    return (
      <ModalContainer title="Delete User Word" toggleModal={toggleModal}>
        <form styleName="modal" action={handleDelete}>
          <p styleName="modal__alert">
            Are you sure you want to delete your words by duration?
          </p>
          <p styleName="modal__message">You can't undo this action.</p>
          <div styleName="modal__buttons">
            <button
              styleName="modal__button modal__button--cancel"
              onClick={handleCancel}
            >
              Cancel
            </button>
            <button
              styleName="modal__button modal__button--delete"
              type="submit"
              disabled={isPending}
            >
              Delete
            </button>
          </div>
        </form>
      </ModalContainer>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);

const USER_WORD_NOTIFICATION_MSGS = {
  deleteUserWord: () => {
    return "You have successfully deleted a word!";
  },
};
