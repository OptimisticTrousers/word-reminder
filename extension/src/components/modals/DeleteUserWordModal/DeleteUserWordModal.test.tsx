import { createRoutesStub, Outlet } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import userEvent from "@testing-library/user-event";

import { DeleteUserWordModal } from "./DeleteUserWordModal";
import * as hooks from "../../../hooks/useNotificationError";
import { wordService } from "../../../services/word_service";

describe("DeleteUserWordReminderModal component", () => {
  const json = {
    word: { id: "1" },
  };

  const testUser = {
    id: "1",
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders the form with alert and message", async () => {
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      wordId: "1",
    };
    const queryClient = new QueryClient();
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteUserWordModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);

    const { asFragment } = render(<Stub initialEntries={["/"]} />);

    const alert = screen.getByText(
      "Are you sure you want to delete your user word?"
    );
    const message = screen.getByText("You can't undo this action.");

    expect(alert).toBeInTheDocument();
    expect(message).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls the functions to delete a word", async () => {
    const queryClient = new QueryClient();
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      wordId: "1",
    };
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteUserWordModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);
    const status = 200;
    const mockDeleteUserWord = vi
      .spyOn(wordService, "deleteUserWord")
      .mockImplementation(async () => {
        return { json, status };
      });
    const user = userEvent.setup();

    render(<Stub initialEntries={["/"]} />);

    const deleteButton = screen.getByRole("button", { name: "Delete" });
    await user.click(deleteButton);
    expect(mockDeleteUserWord).toHaveBeenCalledTimes(1);
    expect(mockDeleteUserWord).toHaveBeenCalledWith({
      userId: testUser.id,
      wordId: props.wordId,
    });
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
  });

  it("calls the functions to show a notification error", async () => {
    const queryClient = new QueryClient();
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      wordId: "1",
    };
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteUserWordModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);
    const status = 200;
    const message = "Error Message.";
    const mockDeleteUserWord = vi
      .spyOn(wordService, "deleteUserWord")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const mockShowNotificationError = vi.fn();
    vi.spyOn(hooks, "useNotificationError").mockImplementation(() => {
      return { showNotificationError: mockShowNotificationError };
    });
    const user = userEvent.setup();

    render(<Stub initialEntries={["/"]} />);

    const deleteButton = screen.getByRole("button", { name: "Delete" });
    await user.click(deleteButton);
    expect(mockDeleteUserWord).toHaveBeenCalledTimes(1);
    expect(mockDeleteUserWord).toHaveBeenCalledWith({
      userId: testUser.id,
      wordId: props.wordId,
    });
    expect(mockShowNotificationError).toHaveBeenCalledTimes(1);
    expect(mockShowNotificationError).toHaveBeenCalledWith({
      json: { message },
      status,
    });
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
  });

  it("disables the delete button when the mutation is loading", async () => {
    const queryClient = new QueryClient();
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      wordId: "1",
    };
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteUserWordModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);
    const status = 200;
    const delay = 50;
    const mockDeleteUserWord = vi
      .spyOn(wordService, "deleteUserWord")
      .mockImplementation(async () => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json, status });
          }, delay);
        });
      });
    const user = userEvent.setup();

    render(<Stub initialEntries={["/"]} />);

    const deleteButton = screen.getByRole("button", { name: "Delete" });
    await user.click(deleteButton);

    expect(deleteButton).toBeDisabled();
    expect(mockDeleteUserWord).toHaveBeenCalledTimes(1);
    expect(mockDeleteUserWord).toHaveBeenCalledWith({
      userId: testUser.id,
      wordId: props.wordId,
    });
  });

  it("calls the functions to close a modal", async () => {
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      wordId: "1",
    };
    const user = userEvent.setup();
    const queryClient = new QueryClient();
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteUserWordModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);
    render(<Stub initialEntries={["/"]} />);

    const cancelButton = screen.getByRole("button", { name: "Cancel" });
    await user.click(cancelButton);

    expect(props.toggleModal).toHaveBeenCalledTimes(1);
    expect(props.toggleModal).toHaveBeenCalledWith();
  });
});
