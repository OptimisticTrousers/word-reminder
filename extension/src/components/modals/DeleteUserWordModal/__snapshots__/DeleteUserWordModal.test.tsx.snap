// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DeleteUserWordReminderModal component > renders the form with alert and message 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="modal-title"
    aria-modal="true"
    class="_modal_811945"
    role="dialog"
  >
    <div
      class="scale-down _modal__container_811945"
    >
      <header
        class="_modal__header_811945"
      >
        <button
          aria-label="Close modal"
          class="_modal__button_811945"
        >
          <svg
            class="lucide lucide-square-x _modal__icon_811945"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="3"
            />
            <path
              d="m15 9-6 6"
            />
            <path
              d="m9 9 6 6"
            />
          </svg>
        </button>
        <div
          class="_modal__title-bar_811945"
          id="modal-title"
        >
          <h2
            class="_modal__title_811945"
          >
            Delete User Word
          </h2>
        </div>
      </header>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="_modal_5625f4"
      >
        <p
          class="_modal__alert_5625f4"
        >
          Are you sure you want to delete your words by duration?
        </p>
        <p
          class="_modal__message_5625f4"
        >
          You can't undo this action.
        </p>
        <div
          class="_modal__buttons_5625f4"
        >
          <button
            class="_modal__button_5625f4 _modal__button--cancel_5625f4"
          >
            Cancel
          </button>
          <button
            class="_modal__button_5625f4 _modal__button--delete_5625f4"
            type="submit"
          >
            Delete
          </button>
        </div>
      </form>
    </div>
  </div>
  ;
</DocumentFragment>
`;
